sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -705,6 +718,17 @@ void reset() throws Exception {
         resetFunction.run();
       }
     }
+
+    void shutdown() {
+      for (ThrowingRunnable tearDownFunction : getTearDownFunctions()) {
+        LOG.debug("Tearing down function {}", tearDownFunction);
+        try {
+          tearDownFunction.run();
+        } catch (Exception e) {
+          LOG.error("Failed to call teardown function: {}", e);
```

Comment:
Note that within this change, exceptions from `shutdown` will not fail the bundle. Alternatively an error will be shown from sdk harness log.

Reply:
Interesting. I did not find good documentation on the expectation about whether a bundle should fail if `TearDown` fails. The most detailed docs are the [javadoc](https://beam.apache.org/releases/javadoc/2.27.0/org/apache/beam/sdk/transforms/DoFn.Teardown.html) and they do not describe this. That seems like we need to describe the expectation.

Note that I think this will call http://www.slf4j.org/apidocs/org/slf4j/Logger.html#error(java.lang.String,java.lang.Throwable) because it is the most precise overloaded method.

I think that is good because it will give a stack trace.

But the `{}` will not be replaced with anything.

Reply:
We can elaborate more on the error message, like saying `There is an exception when calling teardown, but this exception will not fail the execution.`

Reply:
I was thinking somewhere like https://beam.apache.org/documentation/programming-guide/#dofn

Reply:
It seems like dofn lifecycle section is about still non-portable execution. We should add(or update it to) portable execution. I'll add some javadoc to DoFn first and we can update dofn lifecycle together.

Reply:
The document is for a user, who should not have to think about portable execution. The contract for which methods are called should be independent of portable execution. For example some runners may choose to execute some steps directly even if they support portability.

Reply:
> For example some runners may choose to execute some steps directly even if they support portability.
I don't think that's the case. Would you like to share some examples in your mind?

Reply:
I do not mean that it occurs today. I mean that it is permitted.

Reply:
All these steps are defined inside DoFn, right? I would image that sdk harness should be responsible for managing them. Based on the contract, the runner should talk to sdk harness via fnapi. Otherwise, that seems wrong to me.

Reply:
But that is actually a tagent. The important thing is that users can understand how their `DoFn` will be invoked without reasoning about the portability protocols. Like "during a failure, `@Teardown` will be called if it is possible, but not always". We need a spec about what happens when any of the methods fail. I think like "Since `@TearDown` executes independent of any data processing, if `@TearDown` fails, the pipeline is permitted to continue but resources will not be collected."

Reply:
The Fn API is really an implementation detail of this higher-level contract about how we call user functions.